dbms/src/Processors/Formats/Impl/RegexpRowInputFormat.cpp

```diff
@@ -0,0 +1,190 @@
+                break;
+            case FieldFormat::Quoted:
+                if (parse_as_nullable)
+                    read = DataTypeNullable::deserializeTextQuoted(*columns[index], field_buf, format_settings, type);
+                else
+                    type->deserializeAsTextQuoted(*columns[index], field_buf, format_settings);
+                break;
+            case FieldFormat::Csv:
+                if (parse_as_nullable)
+                    read = DataTypeNullable::deserializeTextCSV(*columns[index], field_buf, format_settings, type);
+                else
+                    type->deserializeAsTextCSV(*columns[index], field_buf, format_settings);
+                break;
+            case FieldFormat::Json:
+                if (parse_as_nullable)
+                    read = DataTypeNullable::deserializeTextJSON(*columns[index], field_buf, format_settings, type);
+                else
+                    type->deserializeAsTextJSON(*columns[index], field_buf, format_settings);
+                break;
+        }
+    }
+    catch (Exception & e)
+    {
+        e.addMessage("(while read the value of key " +  getPort().getHeader().getByPosition(index).name + ")");
+        throw;
+    }
+    return read;
+}
+
+void RegexpRowInputFormat::readFieldsFromMatch(MutableColumns & columns, RowReadExtension & ext)
+{
+    if (matched_fields.size() != columns.size())
+        throw Exception("The number of matched fields in line doesn't match the number of columns.", ErrorCodes::INCORRECT_DATA);
+
+    ext.read_columns.assign(columns.size(), false);
+    for (size_t columns_index = 0; columns_index < columns.size(); ++columns_index)
+    {
+        ext.read_columns[columns_index] = readField(columns_index, columns);
+    }
+}
+
+bool RegexpRowInputFormat::readRow(MutableColumns & columns, RowReadExtension & ext)
+{
+    if (buf.eof())
+        return false;
+
+    PeekableReadBufferCheckpoint checkpoint{buf};
+
+    size_t line_size = 0;
+
+    while (!buf.eof() && *buf.position() != '\n' && *buf.position() != '\r')
+    {
+        ++buf.position();
+        ++line_size;
+    }
+
+    buf.makeContinuousMemoryFromCheckpointToPos();
+    buf.rollbackToCheckpoint();
+
+    bool match = RE2::FullMatchN(re2::StringPiece(buf.position(), line_size), regexp, re2_arguments_ptrs.data(), re2_arguments_ptrs.size());
+    bool read_line = true;
+
+    if (!match)
+    {
+        if (!format_settings.regexp.skip_unmatched)
+            throw Exception("Line \"" + std::string(buf.position(), line_size) + "\" doesn't match the regexp.", ErrorCodes::INCORRECT_DATA);
+        read_line = false;
+    }
+
+    if (read_line)
+        readFieldsFromMatch(columns, ext);
+
+    buf.position() += line_size;
+
+    // Two sequential increments are needed to support DOS-style newline ("\r\n").
+    if (!buf.eof() && *buf.position() == '\r')
+        ++buf.position();
+
+    if (!buf.eof() && *buf.position() == '\n')
```

Comment:
The code is slightly incorrect:
it support \r\n, \n and \r.

But correct code should support only \n and \r\n.